File: Creation/Builder/Cpp/builder.cpp
#include <iostream>
#include <string>

// 1. Product
class Car
{
private:
    int wheels;
    std::string color;
    std::string engine;

public:
    void setWheels(int w) { wheels = w; }
    void setColor(const std::string &c) { color = c; }
    void setEngine(const std::string &e) { engine = e; }

    void showDetails() const
    {
        std::cout << "Car with " << wheels << " wheels, color " << color << " and " << engine << " engine." << std::endl;
    }
};

// 2. Builder
class CarBuilder
{
public:
    virtual ~CarBuilder() {}
    virtual void setWheels(int w) = 0;
    virtual void setColor(const std::string &c) = 0;
    virtual void setEngine(const std::string &e) = 0;
    virtual Car *build() = 0;
};

// 3. ConcreteBuilder
class SedanCarBuilder : public CarBuilder
{
private:
    Car car;

public:
    void setWheels(int w) override
    {
        car.setWheels(w);
    }

    void setColor(const std::string &c) override
    {
        car.setColor(c);
    }

    void setEngine(const std::string &e) override
    {
        car.setEngine(e);
    }

    Car *build() override
    {
        return &car;
    }
};

// 4. Director
class ManufacturingDirector
{
private:
    CarBuilder *builder;

public:
    void setBuilder(CarBuilder *b)
    {
        builder = b;
    }

    void construct()
    {
        builder->setWheels(4);
        builder->setColor("Red");
        builder->setEngine("V6");
    }
};

int main()
{
    // Client code
    ManufacturingDirector director;
    SedanCarBuilder sedanBuilder;

    director.setBuilder(&sedanBuilder);
    director.construct();

    Car *car = sedanBuilder.build();
    car->showDetails();

    return 0;
}

File: Creation/Builder/Go/builder.go
package main

import "fmt"

// 1. Product
type Car struct {
	Wheels int
	Color  string
	Engine string
}

func (c *Car) ShowDetails() {
	fmt.Printf("Car with %d wheels, color %s and %s engine.\n", c.Wheels, c.Color, c.Engine)
}

// 2. Builder
type CarBuilder interface {
	SetWheels(int) CarBuilder
	SetColor(string) CarBuilder
	SetEngine(string) CarBuilder
	Build() *Car
}

// 3. ConcreteBuilder
type SedanCarBuilder struct {
	car *Car
}

func NewSedanCarBuilder() *SedanCarBuilder {
	return &SedanCarBuilder{&Car{}}
}

func (b *SedanCarBuilder) SetWheels(wheels int) CarBuilder {
	b.car.Wheels = wheels
	return b
}

func (b *SedanCarBuilder) SetColor(color string) CarBuilder {
	b.car.Color = color
	return b
}

func (b *SedanCarBuilder) SetEngine(engine string) CarBuilder {
	b.car.Engine = engine
	return b
}

func (b *SedanCarBuilder) Build() *Car {
	return b.car
}

// 4. Director
type ManufacturingDirector struct {
	builder CarBuilder
}

func (d *ManufacturingDirector) SetBuilder(b CarBuilder) {
	d.builder = b
}

func (d *ManufacturingDirector) Construct(wheels int, color string, engine string) {
	d.builder.SetWheels(wheels).SetColor(color).SetEngine(engine)
}

func main() {
	// Client code
	director := &ManufacturingDirector{}
	sedanBuilder := NewSedanCarBuilder()

	director.SetBuilder(sedanBuilder)
	director.Construct(4, "Red", "V8")

	car := sedanBuilder.Build()
	car.ShowDetails()
}
